hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/replication/health/ClosedWithUnhealthyReplicasHandler.java

```diff
@@ -0,0 +1,111 @@
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+package org.apache.hadoop.hdds.scm.container.replication.health;
+
+import org.apache.hadoop.hdds.protocol.proto.HddsProtos;
+import org.apache.hadoop.hdds.protocol.proto.StorageContainerDatanodeProtocolProtos.ContainerReplicaProto;
+import org.apache.hadoop.hdds.scm.container.ContainerInfo;
+import org.apache.hadoop.hdds.scm.container.ContainerReplica;
+import org.apache.hadoop.hdds.scm.container.ReplicationManagerReport;
+import org.apache.hadoop.hdds.scm.container.replication.ContainerCheckRequest;
+import org.apache.hadoop.hdds.scm.container.replication.ReplicationManager;
+import org.apache.ratis.protocol.exceptions.NotLeaderException;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.Set;
+
+/**
+ * Handler to process containers EC which are closed but have some replicas that
+ * are unhealthy.
+ */
+public class ClosedWithUnhealthyReplicasHandler extends AbstractCheck {
+
+  public static final Logger LOG =
+      LoggerFactory.getLogger(ClosedWithUnhealthyReplicasHandler.class);
+  private final ReplicationManager replicationManager;
+
+  public ClosedWithUnhealthyReplicasHandler(
+      ReplicationManager replicationManager) {
+    this.replicationManager = replicationManager;
+  }
+
+  /**
+   * Handles a closed EC container with unhealthy replicas. Note that if we
+   * reach here, there is no over, under, or mis replication. This handler
+   * will just send commands to delete the unhealthy replicas.
+   *
+   * <p>
+   * Consider the following set of replicas for a closed EC 3-2 container:
+   * Replica Index 1: Closed
+   * Replica Index 2: Closed
+   * Replica Index 3: Closed replica, Unhealthy replica (2 replicas)
+   * Replica Index 4: Closed
+   * Replica Index 5: Closed
+   *
+   * In this case, the unhealthy replica of index 3 should be deleted. This
+   * is not over replication because that unhealthy replica is not available.
+   * </p>
+   * @param request ContainerCheckRequest object representing the container
+   * @return true if this is a closed EC container with unhealthy replicas,
+   * else false
+   */
+  @Override
+  public boolean handle(ContainerCheckRequest request) {
+    ContainerInfo containerInfo = request.getContainerInfo();
+    if (containerInfo.getReplicationType() != HddsProtos.ReplicationType.EC) {
+      return false;
+    }
+    if (containerInfo.getState() != HddsProtos.LifeCycleState.CLOSED) {
+      return false;
+    }
+
+    Set<ContainerReplica> replicas = request.getContainerReplicas();
+    boolean foundUnhealthy = false;
+    // send delete commands for unhealthy replicas
+    for (ContainerReplica replica : replicas) {
+      if (replica.getState() == ContainerReplicaProto.State.UNHEALTHY) {
```

Comment:
We know that an under or over replicated container should never get here, but if there was some bug that meant it did, we could delete unhealthy replicas that don't have another copy.

Do you think we should have one sanity check here - when we find an unhealthy replica with index N, check there is also another replica which is CLOSED and has index X too? That way, we don't need to do a full under / over replicated check here, but we have some protection against another bug causing under-replicated containers coming here somehow.

Reply:
Yes, I was also thinking about having a check. I've added this check to the handler in the latest commit.